src/modm/driver/temperature/max31855.hpp

```diff
@@ -0,0 +1,104 @@
+// coding: utf-8
+// ----------------------------------------------------------------------------
+/*
+ * Copyright (c) 2022, Sarah Vilete
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_MAX31855_HPP
+#define MODM_MAX31855_HPP
+
+#include <modm/processing/resumable.hpp>
+#include <modm/architecture/interface/spi_device.hpp>
+
+namespace modm
+{
+
+/// @ingroup modm_driver_max31855
+struct max31855
+{
+    /// Thermocouple fault bits
+    enum class
+    Fault : uint8_t
+    {
+        ShortCircuitVcc = Bit2,
+        ShortCircuitGnd = Bit1,
+        OpenCircuit     = Bit0
+    };
+
+    struct modm_packed
+    Data
+    {
+        /// @return value associated with the respective fault
+        Fault getFault()
+        {
+            return static_cast<Fault>(data[3] & 0b111);
+        }
+
+        /// @return the thermocouple temperature scaled according to the device documentation
+        float
+        getThermocoupleTemperature()
+        {
+            // convert raw 14 bit readout in 2's complement to a 16 bit signed
+            const int16_t rawTemp = ((static_cast<int16_t>(data[0]) << 8) | (data[1] & 0b11111100)) / 4;
```

Comment:
This won't work for negative temperatures. For the data value of -1°C from the data sheet this implementation [will give a result of 4095 °C](https://gcc.godbolt.org/z/K7hhcT3h8).

The two's complement conversion is not correct due to [integer type promotion](https://en.cppreference.com/w/cpp/language/implicit_conversion#Integral_promotion). The expression `static_cast<int16_t>(data[0]) << 8` and also the whole value before dividing by 4 has the type `int`. Thus, you'll get a big positive value for negative temperatures.

```suggestion
            const int16_t rawTemp = static_cast<int16_t>((data[0] << 8) | (data[1] & 0b11111100)) / 4;
```